### PR TITLE
Optimized dask datashape detection

### PR DESCRIPTION
**Pull request by philippjfr**
Leads to a 2-4x speedup in categorical aggregations and probably even larger improvements for out-of-core aggregation.

- [x] Fixes https://github.com/bokeh/datashader/issues/633

### CODE REVIEW

**Comment by philippjfr** [comment]
This change does not work for categorical data after all.

**Comment by jcrist** [comment]
This might help explain whats going on: http://dask.pydata.org/en/latest/dataframe-design.html#categoricals

Dask uses an empty pandas dataframe to store dtypes/metadata. For categoricals, we sometimes also store the categories. This is a distinction between "known" and "unknown" categories (see the link above).

In this case you know it's a categorical column, but you don't know what the categories are. The previous behavior was incorrect, because you can't assume that the categories in partition 1 are representative of the categories in all partitions for "unknown" categories.

Instead, when a categorical column is desired, for dask you should do something like:

```
c = 'categorical_column_name'
# Convert unknown categorical column to known categories
if not df[c].cat.known:
    df = df.assign(**{c: df[c].cat.as_known()})
```

Note that converting unknown categories to known categories is expensive, as it involves touching every partition. No strong thoughts if this should be done automatically by datashader if unknown categories are found, or if an informative error should be raised telling the user how to fix this. `df.categorize(columns=[c])` would also be equivalent.

If you know the categories beforehand but dask says they're unknown, you can fix this cheaply by using

```
df[c] = df[c].set_categories(categories)
```

**Comment by philippjfr** [comment]
Thanks @jcrist. Your description matches what I thought about categories, however it does seem like ``.head()`` does infer all the categories despite the fact that the chunk that is actually loaded only contains a single category:

screen shot 2018-08-14 at 10 45 45 pm

Is this maybe specific behavior perhaps due to parquet or is it something we can rely on in general?

**Comment by jcrist** [comment]
> Is this maybe specific behavior perhaps due to parquet or is it something we can rely on in general?

This is parquet files written by fastparquet specific (the categorical data is stored in the parquet metadata, which is not a universal thing to do), and not something generic that you can rely on.

**Comment by philippjfr** [comment]
Okay, thanks. I suppose we could speculatively inspect ``column.head(0)`` of the categorical columns to see if the categories are defined to at least take advantage of that, unless you think there will be a large overhead.

**Comment by jcrist** [comment]
Loading even zero rows loads the whole partition (potentially expensive), and all categorical column partitions will result in categorical data even if the `.cat.categories` isn't exhaustive. You really need to ensure that if you're using a categorical column that the categories in that column are known (`assert df[mycol].cat.known`). Note that `as_known` or `categorize` are no-ops if the categories in that column are already known.

**Comment by philippjfr** [comment]
Thanks, really appreciate your expertise.

**Comment by philippjfr** [comment]
I've opened a separate issue for the categorical issue discussed here since it has considerable performance implications. This PR will instead focus on the optimization of the datashape detection.

File: datashader/core.py [hunk 1]
```diff
@@ -7,10 +7,10 @@
 from xarray import DataArray, Dataset
 from collections import OrderedDict
 
-from .utils import Dispatcher, ngjit, calc_res, calc_bbox, orient_array, compute_coords, get_indices, dshape_from_pandas, dshape_from_dask, categorical_in_dtypes
-from .resampling import resample_2d
+from .utils import Dispatcher, ngjit, calc_res, calc_bbox, orient_array, compute_coords, 
+from .utils import get_indices, dshape_from_pandas, dshape_from_dask
```
Comment (philippjfr) [inline datashader/core.py]: Thanks, forgot to add that to my commit apparently.

File: datashader/core.py [hunk 2]
```diff
@@ -514,7 +514,7 @@ def bypixel(source, canvas, glyph, agg):
         cols_to_keep[glyph.x] = True
         cols_to_keep[glyph.y] = True
         if hasattr(glyph, 'z'):
-            cols_to_keep[glyph.z] = True
+            cols_to_keep[glyph.z[0]] = True
```
Comment (jbednar) [inline datashader/core.py]: Where did this come from?  Seems odd that it shows up in this PR's tests?

Comment (philippjfr) [reply datashader/core.py]: Previously this code was called conditionally, i.e. only for dataframes with categorical columns and the tests weren't testing that.

Comment (jbednar) [reply datashader/core.py]: Ok, thanks.

**Pull request merged**